framework/SwiftPlot/HelperFunction.swift

```diff
@@ -14,41 +14,41 @@ public func getNumberOfDigits(_ n: Float) -> Int{
     return count
 }
 
-public func getMaxX(points p: [Point]) -> Float {
-    var max = p[0].x
-    for index in 1..<p.count {
-        if (p[index].x > max) {
-            max = p[index].x
+public func getMaxX<T>(pairs : [Pair<FloatConvertible,T>]) -> Float {
```

Comment:
Should getMaxX be generic about the X input type? If that's the case, you'll need to change the definition to be T, U where T: FloatConvertible and return type T.

Also, is getMaxX the best name for this? maxX would seem to be simpler, or even `argmax` as used by numpy. Could this be defined as an extension to Array, with an appropriate Element type?

Reply:
@BradLarson we cannot use extensions with the array element as generic types yet.
This is what I tried:
```
extension Array where Element:Pair<FloatConvertible,U> {
    var maxX: Float {
        var max = Float(self[0].x)
        for index in 1..<self.count {
            if (Float(self[index].x) > max) {
                max = Float(self[index].x)
            }
        }
        return max
    }
    var minX: Float {
        var min = Float(self[0].x)
        for index in 1..<self.count {
            if (Float(self[index].x) < min) {
                min = Float(self[index].x)
            }
        }
        return min
    }
}
```
But this doesn't work. So I'm letting it be a function for now.

Reply:
@BradLarson also, will there be a problem with type erasure? 
Let's take LineChart as an example.
We have `var primaryAxis = Axis<FloatConvertible,FloatConvertible>()
This makes the series inside axis as Series<FloatConvertible,FloatConvertible>
and the values inside the series also as Pair<FloatConvertible,FloatConvertible>

Now if I change the definition to T,U with T as FloatConvertible, It gives me this error: 
`error: protocol type 'FloatConvertible' cannot conform to 'FloatConvertible' because only concrete types can conform to protocols`

Now I think, to get past this I'll need to make the LineChart class a generic. This doesn't seem like a very nice idea to me...

Reply:
Don't know if it would help, but here's another take on maxX, relying instead on Comparable to avoid casting to Float for every value, while preserving type information:

```
public func maxX<T, U>(_ values: [Pair<T, U>]) -> T where T: Comparable {
    var max = values[0].x
    for index in 1..<values.count {
        if (values[index].x > max) {
            max = values[index].x
        }
    }
    return max
}
```

I could see an argument for making a plot type generic about the series values it would pass in, allowing for a preservation of types for X and Y values. Some of this might require experimentation, creating cases in your sample applications where you try to pass in various data types and simulate common user interactions with the plots to see what produces the easiest API to use and understand. That's the goal, make this as flexible as possible while hiding as much complexity from end users (developers using this API).